src/drivers/driver_zephyr.c

```diff
@@ -5,64 +5,99 @@
  * This software may be distributed under the terms of the BSD license.
  * See README for more details.
  */
+#include <zephyr/kernel.h>
+
 #include "includes.h"
 #include "utils/common.h"
 #include "eloop.h"
 #include "driver_zephyr.h"
+#include "supp_main.h"
 
 K_MBOX_DEFINE(wpa_supp_mbox);
 #define SCAN_TIMEOUT 30
 
-static void wpa_supp_drv_mbox_msg_handler(void *eloop_ctx,
-					  void *timeout_ctx)
+void wpa_supplicant_event_wrapper(void *ctx,
+				enum wpa_event_type event,
+				union wpa_event_data *data)
+{
+	struct wpa_supplicant *wpa_s = ctx;
+
+	wpa_supplicant_event(ctx, event, data);
+
+	if (wpa_s->interface_removed){
+		return;
+	}
+
+	set_dummy_socket_ready();
+}
+
+static int wpa_supp_drv_mbox_msg_handler(void)
 {
 	struct k_mbox_msg msg;
 	struct zep_wpa_supp_mbox_msg_data mbox_msg_data;
 
-	/* Prepare to receive message */
-	msg.size = sizeof(mbox_msg_data);
-	msg.rx_source_thread = K_ANY;
+	while (1) {
+		/* Prepare to receive message */
+		msg.size = sizeof(mbox_msg_data);
+		msg.rx_source_thread = K_ANY;
 
-	k_mbox_get(&wpa_supp_mbox,
-		   &msg,
-		   &mbox_msg_data,
-		   K_FOREVER);
+		k_mbox_get(&wpa_supp_mbox,
+			&msg,
+			&mbox_msg_data,
+			K_FOREVER);
+		wpa_supplicant_event_wrapper(mbox_msg_data.ctx,
+				mbox_msg_data.event,
+				mbox_msg_data.data);
 
-	mbox_msg_data.cb(mbox_msg_data.ctx,
-			 mbox_msg_data.data,
-			 0);
-}
+		if (mbox_msg_data.data)
+			os_free(mbox_msg_data.data);
+	}
 
+	return 0;
+}
 
-void wpa_supp_event_handler(void *ctx,
+K_THREAD_DEFINE(drv_events_thread,
```

Comment:
I would like to better understand the issue you are trying to solve by adding this thread. I have some suspicion and also concern that this might not be the right way to go.

I think the key to properly passing events from the driver to supplicant is in `epool.c` which, at the moment, isn't thread safe.

I think it's going to be easier to talk offline on teams. Please reach me.

Reply:
Ideally we should follow Linux style and use sockets to communicate with driver, that avoid this extra thread and also the need for dummy socket. This is already planned in the roadmap.

Reply:
The problem I see is that you are manipulating non thread safe common structures from two different threads. My understanding goes like this:

* Supplicant eloop runs in `wpa_s_tid`.
* When an event from the driver comes it is put in the mbox. The `drv_events_thread` kicks-in and in turn runs `wpa_supplicant_event` in it's context. 
* The `wpa_supplicant_event` called from `drv_events_thread` manipulates supplicant internal state but it might happen that this is preempted by `wpa_s_tid` and leads to inconsistent state or crash.

Since you have already introduced a socket to trigger eloop processing why not to actually use it?

* Move the dummy socket logic to `driver_zephyr`.
* Remove the extra thread and mbox structure
* When there is an event to handle simply take `mbox_msg_data` and send it over the socket
* In the eloop socket handler read the data and pass it to `wpa_supplicant_event`. It is going to be processed in the correct context.

You probably don't need to set the socket nonblocking. Seems that if you set CONFIG_ELOOP_SELECT the eloop waits on the socket with a timeout.

Reply:
This is exactly what we had in the plan, extending dummy socket to real socket. In fact once we switch to sockets we don't even know an extra thread `drv_events_thread` (this was introduced to terminate the callstack).

> The wpa_supplicant_event called from drv_events_thread manipulates supplicant internal state but it might happen that this is preempted by wpa_s_tid and leads to inconsistent state or crash.

You are right, but we haven't seen any issue till now because, the `wpa_s_tid` thread blocks till `select` timesout, and by which time `wpa_supplciant_event` is long done, but there is a slight window where it can cause issues.

Reply:
Another reason why I haven't done full socket implementation for events right away is that, it has to work with multiple interfaces (probably with different drivers as well), in Linux this is achieved by netlink sockets, dummy sockets just uses a global UDP socket, so, we need to find a replacement for netlink in Zephyr (Per-interface Unix sockets would have been nice if we had Filesystem access).

Reply:
Right. I think what I propose is a low hanging fruit; Then you could extend. 

Anyway. You are aware of the drawbacks and sdk-hostap is ultimately under your ownership, so I won't block.

Reply:
No, you are right, I have now implemented this and its working fine, it uses a global UDP socket but each message takes in the wpa_supplicant handle, so, it should work with multiple VIFs. We can work on converting UDP->Netlink replacement later.

Reply:
I have rebased as there are 4 commits in this area, so, please check 30159acf4 for this fix.